lib/mayaUsd/render/MaterialXGenOgsXml/ShaderGenUtil.cpp

```diff
@@ -0,0 +1,383 @@
+    // Color at vertices also affect topo, but we have not locked a naming scheme to go from index
+    // based to name based as we did for UV sets. We will mark them as topo-affecting, but there is
+    // nothing we can do to link them correctly to a primvar without specifying a naming scheme.
+    "geomcolor",
+    // Geompropvalue are the best way to reference a primvar by name. The primvar name is
+    // topo-affecting. Note that boolean and string are not supported by the GLSL codegen.
+    "geompropvalue",
+    // Swizzles are inlined into the codegen and affect topology.
+    "swizzle",
+    // Conversion nodes:
+    "convert",
+    // Constants: they get inlined in the source.
+    "constant",
+#if MX_COMBINED_VERSION < 13808
+    // Switch, unless all inputs are connected. Bug was fixed in 1.38.8.
+    "switch",
+#endif
+#if MX_COMBINED_VERSION == 13807
+    // Dot became topological in 1.38.7. Reverted in 1.38.8.
+    // Still topological for filename though.
+    "dot",
+#endif
+};
+} // namespace
+
+const std::string& TopoNeutralGraph::getMaterialName()
+{
+    // A material node is always the first node created and will be named N0
+    static const std::string kMaterialName("N0");
+    return kMaterialName;
+}
+
+TopoNeutralGraph::TopoNeutralGraph(const mx::ElementPtr& material)
+{
+    if (!material) {
+        throw mx::Exception("Invalid material element");
+    }
+    std::string message;
+    if (!material->validate(&message)) {
+        const std::string step("Error in original graph:\n");
+        throw mx::Exception(step + message);
+    }
+    _doc = mx::createDocument();
+
+    auto inputDoc = material->getDocument();
+
+    mx::NodePtr materialNode = material->asA<mx::Node>();
+    if (!materialNode) {
+        // We might handle standalone "Output" element at a later stage
+        throw mx::Exception("Material element is not a node.");
+    }
+
+    std::list<mx::NodePtr> nodesToTraverse;
+    mx::NodePtr            surfaceShader;
+    if (materialNode->getCategory() == SURFACEMATERIAL_CATEGORY) {
+        auto dupMaterial = cloneNode(*materialNode, *_doc);
+        auto surfaceInput = materialNode->getInput(SURFACESHADER_TYPE);
+        if (!surfaceInput || !surfaceInput->getConnectedNode()) {
+            throw mx::Exception("Unconnected material node.");
+        }
+        surfaceShader = surfaceInput->getConnectedNode();
+        if (!surfaceShader) {
+            throw mx::Exception("Unconnected material node.");
+        }
+        auto dupSurfaceShader = cloneNode(*surfaceShader, *_doc);
+        dupMaterial->addInput(SURFACESHADER_TYPE, SURFACESHADER_TYPE)
+            ->setConnectedNode(dupSurfaceShader);
+        nodesToTraverse.push_back(surfaceShader);
+    } else {
+        if (materialNode->getType() != SURFACESHADER_TYPE) {
+            throw mx::Exception("Material shader node is not a surfaceshader.");
+        }
+        auto dupMaterial = _doc->addMaterialNode("N" + std::to_string(_nodeIndex));
```

Comment:
(super minor) - Not sure if it's a touch more intuitive to do _nodeIndex++ here instead of the later ++_nodeIndex. I think I saw this elsewhere as well. Just personal preference perhaps.